---
editor_options: 
  markdown: 
    wrap: 72
output:
  github_document
---

```{r setup, echo = FALSE}
suppressPackageStartupMessages({
  library(readxl);
  library(stringr);
  library(dplyr)
})
```

# Hybrid Methylation Alignment Experiments

Aligning the WGBS sequencing data will be challenging, and there are a
variety of ways that it can be done. This will document the process of
testing various bismark parameters.

Key metrics:

-   Mapping efficiency
-   Methylation context percentages
-   Conversion efficiency based on lambda DNA spike

## Prerequisites

-   The base directory is `$SCRATCH/hybrid_methylation`; from here, it
    is `$BASE_DIR`.
-   Reads should be trimmed and accessible from
    `$BASE_DIR/reads/trimmed`.
-   Singularity images have been configured and aliased.
-   The genome should be setup for bismark and available in
    `$BASE_DIR/genome.sh`.
-   The directory `alignment_experiments` exists in `$BASE_DIR`, with
    the following files inside it. - `functions.sh` -
    `slurm_templates/bismark_align_run.slurm` -
    `slurm_templates/bismark_align_trial.slurm` -
    `scripts/bismark_align_se.sh` - `scripts/bismark_align_pe.sh`
-   Files in the scripts directory should be executable

## Determining Alignment Parameters

First, we'll define this function to set up a new experiment

```{bash align_func, eval = FALSE}
source alignment_experiments/functions.sh
# new_alignment_experiment name "pe" "bismark_parms"

# Alignment experiments w/ default parameters
new_alignment_experiment pe_default "pe" " "
new_alignment_experiment se_default "se" " "
new_alignment_experiment pe_relaxed "pe" "-N 1 --score_min L,0,-0.6"
new_alignment_experiment se_relaxed "se" "-N 1 --score_min L,0,-0.6"
new_alignment_experim ent pe_xrelaxed "pe" "-N 1 --score_min L,0,-0.8"
new_alignment_experiment se_xrelaxed "se" "-N 1 --score_min L,0,-0.8"
new_alignment_experiment pe_local "pe" "-N 1 --local"
new_alignment_experiment se_local "se" "-N 1 --local"
new_alignment_experiment pe_local_relax "pe" "-N 1 --local --score_min G,0,-0.6"
new_alignment_experiment se_local_relax "se" "-N 1 --local --score_min G,0,-0.6"

# Run the alignment trials
run_alignment trial pe_default
run_alignment trial se_default
run_alignment trial pe_relaxed
run_alignment trial se_relaxed
run_alignment trial pe_xrelaxed
run_alignment trial se_xrelaxed
run_alignment trial pe_local
run_alignment trial se_local
run_alignment trial pe_local_relax
run_alignment trial se_local_relax
```

### Experimental Parameters and results

The trial set for each of these is the largest trim file,
`lane1-A2-B_S2_L001_1.trim` (and its partner, for PE).

```{r show_experiment, eval = TRUE, echo = FALSE}
alignment_table = read_excel("alignment_experiments.xlsx", sheet = 1) %>%
  mutate(Time = str_replace_all(Time, c("h " = ":", "m " = ":", "s$" = "")))
alignment_table_cap = "Alignment experiment results for the trail file, lane1-A2-B_S2_L001_1.trim."
opts <- options(knitr.kable.NA = "")


knitr::kable(alignment_table)
```

Based on these results, `pe_xrelaxed` is the best option that doesn't
require `local` (e.g., soft-clipping). I'm hesitant about using that
with WGBS data.

Run the alignment on all of samples with `pe_xrelaxed`:

```{bash, eval = FALSE, echo = TRUE}
# First, modify the run slurm file to have 5 hours, just to be safe
# Then, run the full alignment
run_alignment run pe_xrelaxed

```

## De-duplicate the BAM files

De-duplicate the bam files. We'll be moving to the `pe_xrelaxed`
directory and staying there until otherwise noted.

```{bash, eval = FALSE, echo = TRUE}
cds hybrid_methylation/alignment_experiments/pe_xrelaxed

sbatch slurm/bismark_dedup_run.slurm
```

Let's verify that file sizes have changed:

```{bash, eval = FALSE, echo = TRUE}
du -h bams/*bam # Original bams
# Results:
15G     bams/lane1-A1-A_S1_L001_1.trim_bismark_bt2_pe.bam
22G     bams/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.bam
13G     bams/lane1-A3-C_S3_L001_1.trim_bismark_bt2_pe.bam
17G     bams/lane1-A4-D_S4_L001_1.trim_bismark_bt2_pe.bam
16G     bams/lane1-X1-E_S5_L001_1.trim_bismark_bt2_pe.bam
16G     bams/lane1-X3-F_S6_L001_1.trim_bismark_bt2_pe.bam
12G     bams/lane2-X2-E_S7_L002_1.trim_bismark_bt2_pe.bam
14G     bams/lane2-X4-F_S8_L002_1.trim_bismark_bt2_pe.bam
20G     bams/lane2-X5-A_S9_L002_1.trim_bismark_bt2_pe.bam
16G     bams/lane2-X6-B_S10_L002_1.trim_bismark_bt2_pe.bam
12G     bams/lane2-X7-C_S11_L002_1.trim_bismark_bt2_pe.bam
18G     bams/lane2-X8-D_S12_L002_1.trim_bismark_bt2_pe.bam

du -h bams/dedup/*bam  # de-duplicated
# Results:
11G     bams/dedup/lane1-A1-A_S1_L001_1.trim_bismark_bt2_pe.deduplicated.bam
16G     bams/dedup/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.bam
9.1G    bams/dedup/lane1-A3-C_S3_L001_1.trim_bismark_bt2_pe.deduplicated.bam
13G     bams/dedup/lane1-A4-D_S4_L001_1.trim_bismark_bt2_pe.deduplicated.bam
12G     bams/dedup/lane1-X1-E_S5_L001_1.trim_bismark_bt2_pe.deduplicated.bam
11G     bams/dedup/lane1-X3-F_S6_L001_1.trim_bismark_bt2_pe.deduplicated.bam
7.8G    bams/dedup/lane2-X2-E_S7_L002_1.trim_bismark_bt2_pe.deduplicated.bam
9.6G    bams/dedup/lane2-X4-F_S8_L002_1.trim_bismark_bt2_pe.deduplicated.bam
13G     bams/dedup/lane2-X5-A_S9_L002_1.trim_bismark_bt2_pe.deduplicated.bam
11G     bams/dedup/lane2-X6-B_S10_L002_1.trim_bismark_bt2_pe.deduplicated.bam
8.5G    bams/dedup/lane2-X7-C_S11_L002_1.trim_bismark_bt2_pe.deduplicated.bam
11G     bams/dedup/lane2-X8-D_S12_L002_1.trim_bismark_bt2_pe.deduplicated.bam
```

### Extract Methylation Calls

I tested the methylation extraction script on a single de-duplicated BAM
file (`lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.bam`), then
verify that the output is present.

```{bash, eval = FALSE, echo = TRUE}
# Run extraction trial
sbatch slurm/bismark_extract_trial.slurm
# Wait until complete
du -h methyl_extract/*lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated*
# Results:
25G     methyl_extract/CpG_context_lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.txt
152M    methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.bedGraph.gz
159M    methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.bismark.cov.gz
957M    methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.CpG_report.txt
2.0K    methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.cytosine_context_summary.txt
27K     methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.M-bias.txt
1.0K    methyl_extract/lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated_splitting_report.txt
109G    methyl_extract/Non_CpG_context_lane1-A2-B_S2_L001_1.trim_bismark_bt2_pe.deduplicated.txt

# This is more or less what's expected, so run the rest

sbatch slurm/bismark_extract_run.slurm
```

### Compress & Backup to `WORK`

This has generated a lot of files, so it's worth backing up the
important ones to `WORK` in case `SCRATCH` is purged.

```{bash, eval = FALSE, echo = TRUE}
WRK=${PWD/$SCRATCH/$WORK} # Work version of pe_xrelaxed
mkdir -p $WRK
mkdir -p $WRK/bams/dedup

# Some files need compressing
sbatch slurm/compress_bismark_results.slurm

# Backup reports
cp -r reports $WRK/reports

# Wait on these next two until the slurm job is complete:

# Backup deduplicated BAMS
cp -r bams/dedup/*bam $WRK/bams/dedup/ &

# Backup extracted methylation information
cp -r methyl_extract/*{.gz,_summary.txt,M-bias.txt,_splitting_report.txt} $WRK/methyl_extract
```

# ToDo:

Sanity checks with IGV viewer
Lambda Spike?

SNPsplit
